Cabal/Distribution/PackageDescription/Check.hs

```diff
@@ -245,6 +246,12 @@ checkLibrary pkg lib =
       PackageDistInexcusable $
            "To use the 'required-signatures' field the package needs to specify "
         ++ "at least 'cabal-version: >= 1.21'."
+
+    -- check that all autogen-modules appear on other-modules or exposed-modules
+  , checkVersion [1,25] (not $ and $ map (flip elem (libModules lib)) (libModulesAutogen lib)) $
```

Comment:
I don't think these tests have the intended semantics you want. What you have here is, "If the Cabal version on your file is 1.25 or later, THEN check that every autogen-modules appears in libModules." That means if the Cabal version is lower, this check gets skipped.